File: app/src/main/java/example/com/paycoinqtest/adapter/ReposAdapter.java
package example.com.paycoinqtest.adapter;


import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import example.com.paycoinqtest.R;
import example.com.paycoinqtest.model.RepoInfo;

public class ReposAdapter extends RecyclerView.Adapter<ReposAdapter.ViewHolder> {
	private List<RepoInfo> data = new ArrayList<>();
	private Context context;

	public ReposAdapter(@NonNull Context context, @NonNull List<RepoInfo> data) {
		this.data = data;
		this.context = context;
	}

	@Override
	public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
		View view = LayoutInflater.from(parent.getContext())
				.inflate(R.layout.list_item, parent, false);
		TextView txtName = (TextView) view.findViewById(R.id.txt_name);

		return new ViewHolder(view, txtName);
	}

	@Override
	public void onBindViewHolder(final ViewHolder holder, int position) {
	}

	public void add(RepoInfo newItem) {
		data.add(newItem);
		notifyDataSetChanged();
	}

	@Override
	public int getItemCount() {
		return data.size();
	}

	static class ViewHolder extends RecyclerView.ViewHolder {
		TextView txtName;

		ViewHolder(View itemView, TextView txtName) {
			super(itemView);
			this.txtName = txtName;
		}
	}
}

File: app/src/main/java/example/com/archtest/adapter/ReposAdapter.kt
package example.com.archtest.adapter

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import example.com.archtest.R
import example.com.archtest.data.RepoInfo
import java.util.ArrayList

class ReposAdapter(d: MutableList<RepoInfo>) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    //item types
    private val ITEM = 1
    private val FOOTER = 2

    private var data: MutableList<RepoInfo> = d
    private var isFooterAdded = false

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {

        return when (viewType) {
            ITEM -> createItemViewHolder(parent)
            FOOTER -> createFooterViewHolder(parent)
            else -> createItemViewHolder(parent)
        }
    }

    private fun createItemViewHolder(parent: ViewGroup): ItemViewHolder {
        val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.list_item, parent, false)
        val txtName = view.findViewById<TextView>(R.id.txt_name)

        return ItemViewHolder(view, txtName)
    }

    private fun createFooterViewHolder(parent: ViewGroup): FooterViewHolder {
        val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.list_item_progress, parent, false)

        return FooterViewHolder(view)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (getItemViewType(position)) {
            ITEM -> (holder as ItemViewHolder).txtName.text = data[position].name
            FOOTER -> {
            }
            else -> {
            }
        }
    }

    fun addProgressFooter() {
        if (isFooterAdded) {
            return
        }

        isFooterAdded = true
        add(RepoInfo())
    }

    fun removeProgressFooter() {
        if (!isFooterAdded) {
            return
        }
        isFooterAdded = false

        val position = data.size - 1
        val item = getItem(position)

        if (item != null) {
            data.removeAt(position)
            notifyItemRemoved(position)
        }
    }

    private fun getItem(position: Int): RepoInfo? {
        return data[position]
    }

    override fun getItemViewType(position: Int): Int {
        return if (isLastPosition(position) && isFooterAdded) FOOTER else ITEM
    }

    private fun isLastPosition(position: Int): Boolean {
        return position == data.size - 1
    }

    private fun add(item: RepoInfo) {
        data.add(item)
        notifyItemInserted(data.size - 1)
    }

    fun setData(items: List<RepoInfo>) {
        data = ArrayList(items)
        notifyDataSetChanged()
    }

    override fun getItemCount(): Int {
        return data.size
    }

    private class ItemViewHolder internal constructor(itemView: View, internal var txtName: TextView) : RecyclerView.ViewHolder(itemView)

    private class FooterViewHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView)
}